### PR TITLE
🧙🏻‍♂️ You shall not pass: UI wiring up (dApp permission request)

### LINKED ISSUE #480: dApp permission request UI build and wire up

### PR DESCRIPTION
**Pull request by greg-nagy**
Gandalf shall not pass

# Purpose

This PR will connect the [⚡️ dApp connection approval UI](https://github.com/tallycash/extension/pull/530) into the [dApp permission request flow](https://github.com/tallycash/extension/pull/519).

There is a good chance that this will also include the [⚡️ Connected dApp status UI](https://github.com/tallycash/extension/pull/541)

See #391 
Closes #480

### CODE REVIEW

**Comment by greg-nagy** [comment]
This implementation still has some issues, but it's good enough for review and merge.

File: background/services/provider-bridge/index.ts [hunk 1]
```diff
@@ -59,98 +61,90 @@ export default class ProviderBridgeService extends BaseService<Events> {
...
   }
 
-  onMessageListener(
-    port: Required<browser.Runtime.Port>
-  ): (event: PortRequestEvent) => Promise<void> {
+  async onMessageListener(
+    port: Required<browser.Runtime.Port>,
+    event: PortRequestEvent
+  ): Promise<void> {
     const url = port.sender.url as string
```
[the remark this answers is not in the document]
Comment (Shadowfiend) [reply background/services/provider-bridge/index.ts:78]: Also… We need to manage this by origin. `(new URL(port.sender.url)).origin` should do the trick.

**Review by Shadowfiend — APPROVED** [review]
Left notes. Let's go!

File: background/redux-slices/selectors/providerBridgeSelectors.ts [hunk 2]
```diff
@@ -15,3 +15,10 @@ export const selectPendingPermissionRequests = createSelector(
     return permissionRequests.filter((p) => p.state === "request")
   }
 )
+
+export const perndingPermission = createSelector(
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply background/redux-slices/selectors/providerBridgeSelectors.ts]: ended up with `selectCurrentPendingPermission`

File: background/services/provider-bridge/index.ts [hunk 3]
```diff
@@ -59,98 +61,90 @@ export default class ProviderBridgeService extends BaseService<Events> {
 
     browser.runtime.onConnect.addListener(async (port) => {
       if (port.name === EXTERNAL_PORT_NAME && port.sender?.url) {
-        const listener = this.onMessageListener(
-          port as Required<browser.Runtime.Port>
-        )
-        port.onMessage.addListener(listener)
+        port.onMessage.addListener((event) => {
+          this.onMessageListener(port as Required<browser.Runtime.Port>, event)
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply background/services/provider-bridge/index.ts:65]: I need the port in the `onMessageListener` so I have to pass that in. 

Tried a couple of other solution like checking for port.sender with an if, but did not work. I don't have a better solution, but def open to suggestions.

File: background/services/provider-bridge/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (greg-nagy) [reply background/services/provider-bridge/index.ts:78]: We can not even get here without sender AND url being set `if (port.name === EXTERNAL_PORT_NAME && port.sender?.url) {`

The `!` was a good idea, I used that. The linter does not complain tho.

✅ Also… We need to manage this by origin. (new URL(port.sender.url)).origin should do the trick.

File: background/services/provider-bridge/index.ts [hunk 4]
```diff
@@ -59,98 +61,90 @@ export default class ProviderBridgeService extends BaseService<Events> {
...
+    const title = port.sender.tab?.title ?? ""
 
-    return async (event: PortRequestEvent) => {
-      // a port: browser.Runtime.Port is passed into this function as a 2nd argument by the port.onMessage.addEventListener.
-      // This contradicts the MDN documentation so better not to rely on it.
-      logger.log(
-        `background: request payload: ${JSON.stringify(event.request)}`
-      )
+    // a port: browser.Runtime.Port is passed into this function as a 2nd argument by the port.onMessage.addEventListener.
+    // This contradicts the MDN documentation so better not to rely on it.
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply background/services/provider-bridge/index.ts:83]: CleanShot 2021-12-11 at 15 12 02

Only event and port gets passed into the event listener. 

I believe this footprint belongs to the [one time message passing](https://developer.chrome.com/docs/extensions/mv3/messaging/#simple)

Tried to simplify the code a bit more but could not get any relevant changes 
CleanShot 2021-12-11 at 15 24 42

File: background/services/provider-bridge/index.ts [hunk 5]
```diff
@@ -59,98 +61,90 @@ export default class ProviderBridgeService extends BaseService<Events> {
...
   async requestPermission(permissionRequest: PermissionRequest) {
-    let blockResolve: (value: unknown) => void | undefined
-    const blockUntilUserAction = new Promise((resolve) => {
-      blockResolve = resolve
-    })
-
     this.emitter.emit("permissionRequest", permissionRequest)
-    await ProviderBridgeService.showDappConnectWindow()
+    await ProviderBridgeService.showDappConnectWindow(
+      ALLOWED_QUERY_PARAM_PAGE.dappConnect
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply background/services/provider-bridge/index.ts:122]: I thought about it but did not see any benefit of using it. At runtime it's a bit more confusing to use for me so I did stick with a simple object.

File: provider-bridge-shared/index.ts [hunk 6]
```diff
@@ -104,3 +109,20 @@ export function isWindowResponseEvent(
 export function isPortResponseEvent(arg: unknown): arg is PortResponseEvent {
   return isObject(arg) && isString(arg.id) && !isUndefined(arg.result)
 }
+
+export const ALLOWED_QUERY_PARAM_PAGE = {
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply provider-bridge-shared/index.ts:113]: So there is a disctinction between enums and constants in this codebase? To me they have the same meaning.

File: provider-bridge-shared/index.ts [hunk 7]
```diff
@@ -113,7 +113,7 @@ export function isPortResponseEvent(arg: unknown): arg is PortResponseEvent {
 export const ALLOWED_QUERY_PARAM_PAGE = {
   permissions: "/permission",
```
[the remark this answers is not in the document]
Comment (greg-nagy) [reply provider-bridge-shared/index.ts:114]: nope :)

**Comment by greg-nagy** [comment]
https://github.com/tallycash/extension/pull/570/commits/451a5ea96f8d8650c46ad7a6eccff8c5728347ad made these changes based on this review

**Pull request merged**